src/lib/Embed.hs

```diff
@@ -188,6 +190,28 @@ buildNAbsAux bs body = do
      return (fmap Bind vs, result)
   return (Abs bs' $ wrapDecls decls ans, aux)
 
+buildDataDef :: MonadEmbed m
+             => Name -> Nest Binder -> ([Atom] -> m [DataConDef]) -> m DataDef
+buildDataDef tyConName paramBinders body = do
+  ((paramBinders', dataDefs), _) <- scopedDecls $ do
+     vs <- freshNestedBinders paramBinders
+     result <- body $ map Var $ toList vs
+     return (fmap Bind vs, result)
+  return $ DataDef tyConName paramBinders' dataDefs
+
+buildImplicitNaryLam :: MonadEmbed m => (Nest Binder) -> ([Atom] -> m Atom) -> m Atom
+buildImplicitNaryLam Empty body = body []
+buildImplicitNaryLam (Nest b bs) body =
+  buildLam b ImplicitArrow $ \x -> do
+    bs' <- substEmbed (b@>x) bs
+    buildImplicitNaryLam bs' $ \xs -> body $ x:xs
+
+recGet :: Label -> Atom -> Atom
+recGet l x = do
+  let (RecordTy (Ext r _)) = getType x
```

Comment:
Shouldn't you assert that this has `NoExt`? Otherwise there is no way to know if there won't be more instance of your label coming later? Also, this does retrieve only the first element under a given label, without any assertion that there are no more of those. Is that expected? If so, please reflect that in the name.

Reply:
I'll rename to `recGetHead`. I think it's fine if there are more copies of the same label coming later.